service/subresource_destroyer.go

```diff
@@ -0,0 +1,158 @@
+package service
+
+import (
+	"fmt"
+
+	"github.com/RedHatInsights/sources-api-go/dao"
+	"github.com/RedHatInsights/sources-api-go/kafka"
+	logging "github.com/RedHatInsights/sources-api-go/logger"
+	"github.com/RedHatInsights/sources-api-go/model"
+)
+
+// DeleteCascade removes the resource type and all its dependants, raising an event for every deleted resource. Returns
+// an error when the resources and its dependants could not be successfully removed.
+//
+// - In the case of the resource being a "Source", it removes its applications, its endpoints and its Red Hat Connector
+// connections.
+// - In the other cases, it removes the resource itself.
+//
+// In both cases the authentications are fetched for every single resource and sub resource, and they get deleted in
+// batch.
+//
+// In the case of not being able to delete authentications, we simply log the error and leave them dangling, since we
+// might either have a database or Vault backing our authentications, and we might need manual action to remove the
+// ones that were left undeleted. In the case of a database it is true that it could be wrapped in a transaction, but
+// with Vault that is not possible. What if we start deleting authentications and then some of them error out? We would
+// be able to recover the database data, but some of their authentications would be gone, therefore leaving the
+// resource in an inconsistent state.
+//
+// Plus, it's not the client's fault that we weren't able to delete the authentications, so by taking this approach
+// the clients don't keep an "unremovable" resource because of some failure in deleting the authentications.
+//
+// Finally, those authentications are safely encrypted so they can stay in their datastores until we manually remove
+// them.
+func DeleteCascade(tenantId *int64, resourceType string, resourceId int64, headers []kafka.Header) error {
+	authenticationsDao := dao.GetAuthenticationDao(tenantId)
+	var authentications []model.Authentication
+
+	switch resourceType {
+	case "Source":
+		sourceDao := dao.GetSourceDao(tenantId)
+		applications, endpoints, rhcConnections, source, err := sourceDao.DeleteCascade(resourceId)
+		if err != nil {
+			return fmt.Errorf(`could not completely delete the source: %s`, err)
+		}
+
+		// Raise events for the deleted applications.
+		var appIds []int64
+		for _, app := range applications {
+			appIds = append(appIds, app.ID)
+			err := RaiseEvent("Application.Delete", &app, headers)
```

Comment:
One teeny thing - it has to be `<Resource>.delete`

Reply:
All right, fixed!